[TEP-0091]Update TEP-0091 to implementable

Yongxuanzhang opened the pull request:
This commit converts tep-0091 to implementable by adding more details
into implementation, proposing solutions to undetermined topics.

Yongxuanzhang commented:
cc: @wlynch @dibyom @afrittoli 
Really appreciate your reviews! 🙏

bendory commented:
Can you please clarify in the TEP when verification takes place? Some of my questions that need clarity:

- Is the verification on the `Task`, the `TaskRun`, or both? Examples:
  - When I `kubectl apply` the `Task` definition, is it rejected if not verifiable?
  - There are Tasks where the steps are not resolved until `TaskRun`, and they may even depend on params. For example, the `BASE_IMAGE` param of `git-cli`.
- I suspect that continuous verification is out of scope, but we should say so in the TEP. For example, I `apply` a Task which is verified and signed, but later I learn that the key used to sign it has been compromised, so I decide that I no longer accept that key. With continuous verification, that task would now be blocked without me taking additional action. With a single-shot verification at `apply` time, said task would still be allowed.

Yongxuanzhang commented:
> Can you please clarify in the TEP when verification takes place? Some of my questions that need clarity:
> 
> * Is the verification on the `Task`, the `TaskRun`, or both? Examples:
>   
>   * When I `kubectl apply` the `Task` definition, is it rejected if not verifiable?
>   * There are Tasks where the steps are not resolved until `TaskRun`, and they may even depend on params. For example, the `BASE_IMAGE` param of `git-cli`.
> * I suspect that continuous verification is out of scope, but we should say so in the TEP. For example, I `apply` a Task which is verified and signed, but later I learn that the key used to sign it has been compromised, so I decide that I no longer accept that key. With continuous verification, that task would now be blocked without me taking additional action. With a single-shot verification at `apply` time, said task would still be allowed.

Sure! 
For first question, it will happen on both, so if user apply a task to cluster, it will be verified. And later when a taskrun wants to refer to this Task, we will fetch this task to verify again. 
For params we don't verify on the resolved params, within the scope of this TEP we only secure the Task/Pipeline. 

For second question, since we also do verification on the tasks when applying taskrun&pipelinerun, if the key is invalid then it will be rejected. That's why we need to do it for both.

bendory commented:
> For second question, since we also do verification on the tasks when applying taskrun&pipelinerun, if the key is invalid then it will be rejected. That's why we need to do it for both. Not sure if this could cover the issue you described.

So you are saying that at the time of `taskRun` (or `pipelineRun`), we re-verify the `Task` / `Pipeline`?

(IMO, the questions I'm asking need to be clarified in the TEP itself, though I understand that may mean addressing them in these comments threads first.)

Yongxuanzhang commented:
> > For second question, since we also do verification on the tasks when applying taskrun&pipelinerun, if the key is invalid then it will be rejected. That's why we need to do it for both. Not sure if this could cover the issue you described.
> 
> So you are saying that at the time of `taskRun` (or `pipelineRun`), we re-verify the `Task` / `Pipeline`?
> 
> (IMO, the questions I'm asking need to be clarified in the TEP itself, though I understand that may mean addressing them in these comments threads first.)

Yes, that is the solution we proposed. There are some places in the TEP mentioned this.
But you're right that they are so clear to readers. I will modify and highlight this in the TEP.

bendory commented:
> > > For second question, since we also do verification on the tasks when applying taskrun&pipelinerun, if the key is invalid then it will be rejected. That's why we need to do it for both. Not sure if this could cover the issue you described.
> > 
> > 
> > So you are saying that at the time of `taskRun` (or `pipelineRun`), we re-verify the `Task` / `Pipeline`?
> > (IMO, the questions I'm asking need to be clarified in the TEP itself, though I understand that may mean addressing them in these comments threads first.)
> 
> Yes, that is the solution we proposed. There are some places in the TEP mentioned this. But you're right that they are not so clear to readers. I will modify and highlight this in the TEP.

SGTM -- for the record, it may have just been unclear to _this particular_ reader who is still coming up to speed on everything happening here. 🙃

Yongxuanzhang commented:
> > > > For second question, since we also do verification on the tasks when applying taskrun&pipelinerun, if the key is invalid then it will be rejected. That's why we need to do it for both. Not sure if this could cover the issue you described.
> > > 
> > > 
> > > So you are saying that at the time of `taskRun` (or `pipelineRun`), we re-verify the `Task` / `Pipeline`?
> > > (IMO, the questions I'm asking need to be clarified in the TEP itself, though I understand that may mean addressing them in these comments threads first.)
> > 
> > 
> > Yes, that is the solution we proposed. There are some places in the TEP mentioned this. But you're right that they are not so clear to readers. I will modify and highlight this in the TEP.
> 
> SGTM -- for the record, it may have just been unclear to _this particular_ reader who is still coming up to speed on everything happening here. 🙃

Sorry I didn't mean that. 😂 This is for sure one part not clearly discussed and highlighted. I took it for granted and missed it.

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -62,7 +62,7 @@ attestation claims that are made.
   within the build pipeline itself.
 - Leveraging the above, provide mechanisms to establish a verifiable corpus
-  of community-provided Tasks/Pipelines, Task/Pipelines Bundles and other Tekton types as well in the future.
+  of community-provided Tasks/Pipelines, Task/Pipelines Bundles and other Tekton types as well in the future. They are considered as Tekton resources we need to verify withon the context of this TEP.
```
typo: `withon` --> `within`

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -124,6 +124,8 @@ sign a given OCI image or other artifacts including binaries, scripts etc. This
 This proposal will introduce new verification into Tekton Pipelines' webhook to verify a
 Tekton Resource is tampered or not, and can indicate that the verification must occur.
 
+**Note:** For local resource, it will be verified first when applied to cluster, and when it is referred in taskrun/pipelinerun, the resource will be resolved and verified again to make sure it can still pass the verification.
```
Clarity suggestion: 

Resources (`Task`, `Pipeline`) will be verified *both* when applied to the cluster *and* when referenced in `taskRun`/`pipelineRun`. When the `Run` is created, referenced resources will be resolved and verified again to confirm continued verification. This will prevent the `Run` of a resource that was verifiable when created but is no longer verifiable, perhaps due to key revocation, a security breach, or discovery of a new vulnerability since the time of the initial verification.

Yongxuanzhang replied:
Thank you!!!

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -180,15 +181,16 @@ Task would still be considered verified.
 
 Situation A: Users want to use their own resources and make sure they are not tampered before using in Tekton. The user will need to choose a certain key tool (In the scope of this TEP we will leverage [sigstore](https://github.com/sigstore) to support different keys like cosign or KMS keys) and generate private&public key pair. The user will use the signing cli we provided to sign the yaml files ...
-The public key can be configured from Tekton Pipeline's configmap or in the taskrun/pipelinerun.
+The public key can be configured at Tekton Pipeline.
```
Can you clarify what `at Tekton Pipeline` means? As written, this could mean "as part of the installation of the Pipeline CRD" or "at the time of the installation of this particular `Pipeline` resource"?

Yongxuanzhang replied:
Oh yes, I will clarify this.  It should be "as part of the installation of the Pipeline CRD", like the configmap.

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -298,6 +301,12 @@ spec:
+`keyref` defines the path of public key of this taskRef. All the tasks within a pipeline need to be signed off, and they may come from different sources with different public keys.
+So a `keyref` is needed when signing the pipeline. The signing cli doesn't need to verify the referred keys when signing pipeline.
+
+In this TEP we proposed to choose option 1, for option 2 it would cause a bad user experience that if one pipeline contains a bunch of tasks, then users will need to configure them manually one by one. This would be a big cost for using this feature.
```
I'm fine with making an "out-of-scope" declaration in this `TEP`, but we definitely will want to be able to support multiple keys in the future.

CUJ: "I a cluster admin, and I allow `Pipeline` creation for my application developers only of pipelines that use any of the following list of keys."

Yongxuanzhang replied:
Option 1 also supports multiple keys, I will clarify it in this part！

bendory reviewed (approved):
/lgtm

See clarifications, a typo, and a suggestion.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -141,25 +137,37 @@ In later iterations, this can be extended such that tag-referenced images can
+**Pros vs Cons:**
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Call setdefaults when signing | Help users add missing values |  Need to update signature when there is update in mutating
```
An issue I can see with this is it effectively ties client versions to server versions. This could cause conflicts with API compatibility since any change to SetDefaults that would cause a change in the signature would now be a backwards incompatible change.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +263,64 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+| Config at install level (e.g. configmap) | Easy to config at a central place |  ConfigMap doesn't support list, may need to parse the string or create a new cluster level config to store the keys
+| Add keyref in taskRef&pipelineRef | Each resource can speficy the corresponding key, no need to loop all available keys for one resource's verification| Introduce cost for users to write and maintain YAML files
+
+In this TEP we proposed to choose option 1, for option 2 it would cause a bad user experience that if one pipeline contains a bunch of tasks, then users will need to configure them manually one by one. This would be a big cost for using this feature.
```
(I'd also do this with the options presented in Risks and Mitigations)

Yongxuanzhang replied:
Oh sure!

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +263,64 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+**Pros vs Cons:**
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Config at install level (e.g. configmap) | Easy to config at a central place |  ConfigMap doesn't support list, may need to parse the string or create a new cluster level config to store the keys
```
Yeah I wouldn't recommend a ConfigMap for this - you probably want a dedicated CRD so you can validate the spec, have dedicated RBAC policies, etc.


https://github.com/sigstore/policy-controller and the related [ClusterImagePolicy](https://github.com/sigstore/policy-controller/blob/main/config/300-clusterimagepolicy.yaml) can probably be a good source of inspiration here for configuration of multiple keys. Kyverno / OPA may also be other good places to look for inspiration for image policies.

Yongxuanzhang replied:
Sure. I will update this!

dibyom commented:
/kind tep
/assign

jerop commented:
/assign @wlynch

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -141,25 +137,37 @@ In later iterations, this can be extended such that tag-referenced images can
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Call setdefaults when signing | Help users add missing values |  Need to update signature when there is update in mutating
+| Skip mutating webhook | Address the mutating issue, can also help to detect malicious mutating webhook| Extra cost for Tekton Pipeline developers
```
yeah I imagine it would be "the work needed to change the existing webhook behavior",

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -141,25 +137,37 @@ In later iterations, this can be extended such that tag-referenced images can
+In this TEP we proposed to proceed with option 1 and promo the option 2. An temporary solution for 2 is that we can have a specific feature flag for trusted resources and use it to skip the mutating.
+Options 2,3 and 4 are functionaly equal,  i.e. make sure that when doing verification the resource is not mutated by mutating webhook.
+
+2 is a promising option and can also benefit TEP-0089. But it requres community to reach a consensus, and may introduce development cost like fixing validaing webhook and make sure the mutating functions are invoked in reconciler. In current code base the mutating functions are called after we resolve the local/remote resources. Another benefit of this option is that we can prevent attack from ma ...
```
Oh yes, I will remove the "the mutating functions are invoked in reconciler"

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +263,64 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+| Option | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Config at install level (e.g. configmap) | Easy to config at a central place |  ConfigMap doesn't support list, may need to parse the string or create a new cluster level config to store the keys
+| Add keyref in taskRef&pipelineRef | Each resource can speficy the corresponding key, no need to loop all available keys for one resource's verification| Introduce cost for users to write and maintain YAML files
```
The keyref is not going together with signature so it may not be an issue?
For example, is keyref is configured at taskrun and signature is at task. The attacker doesn't access to taskrun

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -141,25 +137,37 @@ In later iterations, this can be extended such that tag-referenced images can
+| Store the object at mutating webhook before mutating| Fix the issue without skipping the mutating webhook |  May not work if the attck happens in mutating webhook and only modify the spec
+| Verify at mutating webhook | No need to skip the mutating or store in annotation | Doesn't work if there is a malicious mutating webhook after the normal mutating webhook
+
+In this TEP we proposed to proceed with option 1 and promo the option 2. An temporary solution for 2 is that we can have a specific feature flag for trusted resources and use it to skip the mutating.
```
I will change this part, maybe choose option 1 and 2?

pritidesai commented:
S3C WG - waiting for approval

/cc @tektoncd/core-maintainers

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -142,29 +142,9 @@ this will be left out of the scope of this TEP.
 * Mutating webhook is a risk of the proposal. The proposed verification will happen after the mutating webhooks, so the content of a Resource may be mutated and fail the verification. This would be an issue for local cluster resources. And it also means that if we don't address this issue, all signed resources need to be signed again everytime we change the mutating webhook.
 
-  Some possible mitigations/solutions include the following:
+  This TEP proposed to call `SetDefaults` when signing the resources and skip the mutating webhooks when trusted resource feature is enabled.
```
grammar nit: "...when **the** trusted resource..."

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -265,61 +245,7 @@ code.
-In this TEP we proposed to choose option 1, for option 2 it would cause a bad user experience that if one pipeline contains a bunch of tasks, then users will need to configure them manually one by one. This would be a big cost for using this feature.
-
-For option 1, we mainly want to specify a list of trusted keys in Tekton Pipeline, and when verifying we can loop the keys to see if the resource can pass any of these keys.
+In this TEP we proposed to choose store the public keys at installation-level. For example, the keys can be configured in a new CRD and deployed at cluster. This can help to support multiple keys, validate the spec, and have dedicated RBAC policies.
```
grammar nit: "...we propose to store the public keys..."

Yongxuanzhang replied:
Thank you! I will change all other places as well

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -336,39 +262,28 @@ metadata:
-if verification fails.
 
-`skip-verification`. (Optional, default `true`), if true then directly return nil, if false then do the verfication
+`skip-verification`. (Optional, default `true`), if true then directly return nil, if false then do the verification
```
API clarity:
IMO, this option should be `verify` with a default value of `false`. I find it easier to reason about `verify=true` / `verify=false` than to deal with the double-negative of `skip-verification=false`.

Yongxuanzhang replied:
yeah sure!! I want to change it to what you suggested yesterday but somehow forgot it. 😂

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -336,39 +262,28 @@ metadata:
 `kms-pubkey-path`.  (Optional, default empty), it specifies the KMS reference
 
-The key path configuration can be used as default keys when no keys are founded in `keyref`.
+The key path configuration can be used as default keys when no keys are found in `keyref`.
```
IIUC, we are allowing multiple keys to be specified, is this correct?

If so, perhaps we want a single array value rather than separate configurations for `cosign` and `kms`?

```
public-key-paths:
  - '/etc/signing-secrets/cosign.pub'
  - 'gcpkms://<key-path>'
  - 'amskms://<key-path>'
```

This admittedly makes it nominally more difficult to _augment_ the default single value by adding additional keys (you have to specify both the default and additional values), but I think it (1) makes the API more readable and (2) provides a better path for future growth of additional keys / key types.

Yongxuanzhang replied:
Originally I want to use `configmap` as a default place to store public keys, since `configmap data` is only string-string map it doesn't support array. So multiple keys will not be supported. That's why in this tep we suggest using a CRD to handle multiple keys.

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -336,39 +262,28 @@ metadata:
 the remote image and expand it to a Task.
 
-There are several ways to integrate the Trusted Tesource with Remote Resource Resolution
+This TEP proposed to create the remote resolution request before verification in webhook. All alternatives will be discussed at [alternatives](#alternatives).
```
Grammar nit: "...proposes creating the remote resolution request before verification in the webhook."

Second nit: "Alternatives are discussed at..." (I doubt it's ever **all** alternatives! 🙃)

Yongxuanzhang replied:
yes! Thank you! 😄

bendory commented on teps/0091-trusted-resources.md:
```diff
@@ -394,6 +309,104 @@ What other approaches did you consider and why did you rule them out?  These do
+
+### Alternatives for storing the public keys
+
+Option 1: Configure the keys in a dedicated CRD. This can help to support multiple keys in a central place.
```
This is an interesting idea! Can you explain more clearly why it was rejected?

In particular: having keys installed in a separate CRD enables separation of authority -- security folks can centrally control which keys are valid for use and `task` / `pipeline` authors incorporate references to said keys. Now when a key needs to be rotated, the rotation happens in a single place (the CRD) rather than in every `task` / `pipeline` that is configured to use the kay.

Yongxuanzhang replied:
Oh this is not rejected, this is the option we propose in this TEP. But I list all options in the Alternatives section for better comparison. This is a idea suggested by Billy and inspired by  https://github.com/sigstore/policy-controller. 

I will add more details as you suggest in this PR! Thanks!

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -254,47 +276,52 @@ metadata:
-| Fetch remote resources in validating Webhook | Verification fail fast in webhook | Duplicate work for Resolution
-| Verify in Controller | No duplicate work for Resolution | The verification cannot fail fast in webhook. The resources may have been stored in etcd and used by other components
-| Verify in Remote Resolution | No duplicate work for Resolution | Verification coupled with Resolution
+## Threat Models
```
I add this section and want to discuss how this TEP can help to secure the potential threats. 
Please take a look and let me know if they make sense. Thanks!
@wlynch @dibyom

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -313,6 +342,104 @@ What other approaches did you consider and why did you rule them out?  These do
+
+In this TEP we propose to proceed with option 1 and together with option 2.
+
+Option 1 can help to add missing values when users creating resources yaml files.
```
I think this is probably more trouble than it's worth.

In order for this to work, client versions need to stay in sync with the server (and potentially also other server-side configuration like the alpha feature flag). Any time either of these change, you risk breaking the signatures which would be a backwards incompatible change, which can also have rippling effects for the catalog, etc.

If we're skipping the mutating webhook, this already means that we're not modifying the object from what the user already gave us - instead of requiring a replay of the defaulting behavior we can just validate the message as the user gave it to us, and handling the defaulting at reconcile.

Yongxuanzhang replied:
Sure that makes sense! I will remove this.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -141,25 +135,17 @@ In later iterations, this can be extended such that tag-referenced images can
+* Mutating webhook is a risk of the proposal. The proposed verification will happen after the mutating webhooks, so the content of a Resource may be mutated and fail the verification. This would be an issue for local cluster resources. And it also means that if we don't address this issue, all signed resources need to be signed again everytime we change the mutating webhook.
 
-  Some possible mitigations include the following:
+  This TEP proposes to call `SetDefaults` when signing the resources and skip the mutating webhooks when the trusted resource feature is enabled.
```
A caveat to this is skipping the mutating webhook would technically be a backwards incompatible change since it alters the response back to clients (though I highly doubt users are relying on its behavior - AFAICT it only sets param/result types).

Fine while it's opt-in, but if we do want to make the 2 paths consistent (which I'm in favor of) we should plan to start those deprecation timers ASAP.

Yongxuanzhang replied:
We just want to skip the mutating for tasks and pipelines, for taskrun and pipelinerun we can still have the mutating webhook. So removing it seems not ok?

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +162,29 @@ Task would still be considered verified.
-For remote Resources such as OCI bundle, it will take time to fetch the resources. Resources that are not
-configured to be verified will have no change in performance.
+**Verify the Resource via validating webhook.**
+The verification will be done in pipeline's admission webhook, by default the webhook will skip a Resource's verification. This can be configured from configmap by users to decide whether skip the verification or not.
```
| by default the webhook will skip a Resource's verification

Clarification - this is only the signature verification, right?

Might also be worth mentioning warn mode here as well.

Yongxuanzhang replied:
Yes, this is only the signature verification.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +162,29 @@ Task would still be considered verified.
+**Verify the Resource via validating webhook.**
+The verification will be done in pipeline's admission webhook, by default the webhook will skip a Resource's verification. This can be configured from configmap by users to decide whether skip the verification or not.
+
+During verification, the signature will be extracted from the resource, and will first check if public key exsits in the resource's taskrun/pipelinerun. If not then go to config map to use the default keys.
```
| and will first check if public key exsits in the resource's taskrun/pipelinerun

This is no longer true, right?

Yongxuanzhang replied:
Yes, fixed!

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -254,47 +278,52 @@ metadata:
-if verification fails.
 
-`skip-verification`. (Optional, default `true`), if true then directly return nil, if false then do the verfication
+`verify`. (Optional, default `false`), if false then directly return nil, if true then do the verification
```
We'll have more flexibility if this is an enum rather than a bool. I like [Pod Security Admission's policy modes](https://kubernetes.io/docs/concepts/security/pod-security-admission/#pod-security-admission-labels-for-namespaces) of enforce/warn/audit - maybe this can look like:

- No policy = nothing (don't enforce)
- Policy (no mode set) = enforce
- Policy (mode = enforce/warn/audit) = use the given mode

Yongxuanzhang replied:
I think we need a mode to directly skip the whole verification? So maybe (enforce/warn/skip)?

Yongxuanzhang replied:
Oh I get it after reading the following comments. No policy applied indicates the `skip`

wlynch reviewed (changes requested):
Excited for this! Left a bunch of comments digging into some of the gritty details since this is moving to implementable.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -239,6 +243,26 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+
+Example of the Key CRD:
+```yaml
+apiVersion: tekton.dev/v1beta1
````
I remember we won't use v1alpha1 in the future? And we will maintain v1beta1 and v1.

wlynch replied:
We're removing Core Pipeline types from v1alpha1 in favor of v1beta1, but for new types I think we'd want to start from alpha.

Yongxuanzhang replied:
Ok! I will update this, np

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -239,6 +243,26 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+Example of the Key CRD:
+```yaml
+apiVersion: tekton.dev/v1beta1
+Kind: PublicKeys
````
Sure! That sounds good!

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +243,26 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+apiVersion: tekton.dev/v1beta1
+Kind: PublicKeys
+metadata:
+ name: public-keys
```
Sure, I will add namespace here

Yongxuanzhang replied:
I have updated the example and added api by referring to ClusterImagePolicy, I changed the ImagePattern to ResourcePattern, where we can match the resource name and the pattern to support the mapping from resources -> keys

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +243,26 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+metadata:
+ name: public-keys
+spec:
+   OfficialKeys:
+    - keypath1
+    - keypath2
+   ThirdPartyKeys:
+    - keypath1
+    - keypath2
```
Removed this part. 
The new API should support it?

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -239,6 +243,26 @@ The signed Resource can be installed directly on the kubernetes cluster or built
+ name: public-keys
+spec:
+   OfficialKeys:
+    - keypath1
```
added example for keys format

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
+}
+
+type VerificationPolicySpec struct {
+	// Resources defines the patterns of Resources names that should be subject to this policy. For example, we may want to apply this Policy only from a certain github repo. Then the ResourcesPattern should include the path.
```
This could use a bit more details - is this [the CRD resource itself](https://github.com/tektoncd/pipeline/blob/2704d7fb38679d33f2e186a4f9b630482d53c79d/config/300-taskrun.yaml#L15-L18), an [instance of a CRD](https://github.com/tektoncd/pipeline/blob/main/examples/v1beta1/taskruns/task-result.yaml), or something else?

Examples would help here 🙏

Yongxuanzhang replied:
https://github.com/tektoncd/resolution/tree/main/gitresolver#parameters
https://github.com/tektoncd/resolution/tree/main/bundleresolver#parameters
https://github.com/tektoncd/resolution/tree/main/hubresolver#parameters
I want to use the `ResourcesPattern` to enable users to only verify resources from certain source or with certain name. 
For example, the `catalog` for hubresolver, `bundle` for bundleresolver, and use regex to match these fields.

Yongxuanzhang replied:
I'm thinking about a case where User A has a catalog repo, and he somehow exposes his key for that repo. Then we can configure the key for all resources coming from that repo

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
+	Authorities []Authority `json:"authorities"`
+}
+
+type ResourcesPattern struct {
```
nit:
```suggestion
type ResourcePattern struct {
```

(ResourcePatterns sounds better than ResourcesPatterns as a plural)

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
+
+type ResourcesPattern struct {
+	// Pattern defines a resource pattern.
+	Pattern string `json:"glob"`
```
1. Generally try and have the JSON keys match the field name
2. What kind of pattern should this be- glob, regex, etc?

Yongxuanzhang replied:
Oh thanks! I will update this.
I'm thinking of regex match

Yongxuanzhang replied:
Added more details in the docstring

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
+	Keyless *KeylessRef `json:"keyless,omitempty"`
+	// Sources sets the configuration to specify the sources from where to consume the signatures if the signature is not stored in the resource.
+	// +optional
+	Sources []Source `json:"source,omitempty"`
```
Remote signatures would be interesting to support - do you want to include it in this TEP or hold off for another?

dibyom replied:
+1 to adding this later

Yongxuanzhang replied:
yes, hold off for another

wlynch replied:
sg - let's remove the field from here then.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
-  onError: "stopAndFail"
-  cosign-pubkey-path: "/etc/signing-secrets/cosign.pub"
-  kms-pubkey-path: "gcpkms://projects/<project>/locations/<location>/keyRings/<keyring>/cryptoKeys/<key>"
+  resources-verify: true
```
I don't think you need feature flag, assuming that the default behavior of the cluster remains the same if there is no policy defined.

Creating the policy is effectively enabling the feature.

wlynch replied:
oh, I was assuming they wouldn't be marked as generic invalid - it'd be great to have a different error state with an actionable message like `VerificationFailed: denied by policy default/verification-policy` so it's not a complete surprise why a Run is newly failing.

Because adding the policy config is already an opt-in action, also requiring a feature flag seemed redundant to me. But if you want to include it that's fine too!

dibyom commented on teps/0091-trusted-resources.md:
```diff
@@ -313,6 +395,102 @@ What other approaches did you consider and why did you rule them out?  These do
+| Store the object at mutating webhook before mutating| Fix the issue without skipping the mutating webhook |  May not work if the attack happens in mutating webhook and only modify the spec
+| Verify at mutating webhook | No need to skip the mutating or store in annotation | Doesn't work if there is a malicious mutating webhook after the normal mutating webhook
+
+In this TEP we propose to proceed with with option 2.
```
SGTM - are we disabling mutating webhooks all the time or only when this feature is enabled?

Yongxuanzhang replied:
Only when this feature is enabled. And this will only be set for tasks and pipelines not the runs. This is mentioned in the next paragraph

dibyom commented on teps/0091-trusted-resources.md:
```diff
@@ -313,6 +395,102 @@ What other approaches did you consider and why did you rule them out?  These do
+| Verify in Controller after resolving resources| No duplicate work for Resolution | The verification cannot fail fast in webhook. The resources may have been stored in etcd and used by other components
+| Verify in Remote Resolution | No duplicate work for Resolution | Verification coupled with Resolution
+
+In this TEP we propose to fetch remote resources in validating Webhook considering that untrusted resources shouldn't be cond in etcd. And the performance influence of duplicate resolution can be ignored.
```
The way remote resolution works, you'll have to create a resolution request crd from the webhook, wait for the resolver to do its thing and update the status of the resolution request with the pipeline content and then verify. Essentially you'll be implementing a reconciler in the webhook itself :) I don't think we should do that. 

What is the issue with creating a run object that contains a reference to a potentially untrusted resource and storing it in etcd?

Yongxuanzhang replied:
Oh sorry I forgot to update this part. We have reached to an agreement to do it in the reconciler and not in the webhook. So this won't be a problem.

dibyom commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
+
+Before the verification, the signature is extracted from the resource. For the public key there are several options we can choose by configuration.
+
+In this TEP we propose to store the public keys on the Kubernetes installation. For example, the keys can be configured in a new CRD and deployed to the cluster. This can help to support multiple keys, validate the spec, and have dedicated RBAC policies, so no other components in the cluster can modify the keys.
```
I think a CRD makes sense but we'd want to flesh out what the CRD exactly looks like. I'd like to start with the bare-minimum we need i.e. a list of authorities and maybe the patterns to match resources against.
(Would also be useful to call out why a config map would not work for this)

Yongxuanzhang replied:
Sure! I will update the example of this CRD to have more details.

There are some reasons CRD may be better for configmap:
1. In CRD we can configure multiple keys easily (like in the authorities list) but for configmap there is no list, and users need to figure out what will be the key value in this map. 
2. With CRD we can support other features, like mapping keys to the resources via ResourcePattern.  
3. Better support for keyless.

But yes it is no wrong to use configmap. I will add this back as the default choice.

dibyom reviewed (commented):
Thanks for the updates @Yongxuanzhang 
To me, the following seems ok
1. Verifying tasks/pipelines in a validation webhook on creation/updates
2. Verifying taskruns/pipelineruns as the first step in the reconciliation process after a pipeline has been resolved
3. A CRD that stores information on trusted keys to use to verify some subset resources - I'm not 100% sure why we cannot start with a config-map first here but a CRD just to store data seems fine.

What I have a concern with is trying to implement a reconciler in the webhook in order to verify taskruns/pipelineruns on creation. I don't think we should do this.

To merge this as implementable, I also think we need to clearly call out the new CRD being proposed in the Proposal section and provide details on what this CRD will look like and how it will be used in the verification process.

Yongxuanzhang commented:
> Thanks for the updates @Yongxuanzhang To me, the following seems ok
> 
> 1. Verifying tasks/pipelines in a validation webhook on creation/updates
> 2. Verifying taskruns/pipelineruns as the first step in the reconciliation process after a pipeline has been resolved
> 3. A CRD that stores information on trusted keys to use to verify some subset resources - I'm not 100% sure why we cannot start with a config-map first here but a CRD just to store data seems fine.
> 
> What I have a concern with is trying to implement a reconciler in the webhook in order to verify taskruns/pipelineruns on creation. I don't think we should do this.
> 
> To merge this as implementable, I also think we need to clearly call out the new CRD being proposed in the Proposal section and provide details on what this CRD will look like and how it will be used in the verification process.

Yes, in the latest commit I have changed the verification from webhook to reconciler. And after discussing this in pipeline wg, the community lean to do the verification in reconciler not in the webhook. So we don't have to worry about adding a new reconciler logic.

I have also added the new CRD apis in the latest commit with examples.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +161,27 @@ Task would still be considered verified.
 A Resource Author may want their Resource and its
 outputs to be trusted so that provenance attestations can be confidently made. The author will use their own private key to sign the Resource and signatures are stored in `Annotations` map of the Resource.
-The private keys are generated, used and stored by the author. Public keys then are stored as secrets or via URI stored in the configmap. Similar mechanism is used in [Chains](https://github.com/tektoncd/chains).
+The private keys are generated, used and stored by the author. Public keys then are stored at a CRD (Custom Resource Definition).
```
sure, it is included in the api section

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +161,27 @@ Task would still be considered verified.
 
-### Performance (optional)
+**Verify the Resource in reconciler.**
+The verification will be done in pipeline's reconciler, by default we will skip a Resource's signature verification. This can be configured from configmap by users to decide whether to skip the verification or not.
```
Thanks!

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +161,27 @@ Task would still be considered verified.
-available data so there will be no change to performance.
+### Performance
 
+The verification should not introduce too much latency into the reconciler. If the TaskRun/PipelineRun status is marked as verified we should skip the verification.
```
I'm thinking of adding this in reconciler, if we have verified a taskrun/pipelinerun referred resources and mark them as verified, we don't have to call verification again.

Yongxuanzhang replied:
Or we need to do it for each reconcile?

Yongxuanzhang replied:
I will remove this for now. I think it is more secure to verify for each reconcilation.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -176,28 +161,27 @@ Task would still be considered verified.
-available data so there will be no change to performance.
+### Performance
 
+The verification should not introduce too much latency into the reconciler. If the TaskRun/PipelineRun status is marked as verified we should skip the verification.
```
Sure! Mainly for 2 reasons, we decided not to verify at webhook, so there is no extra cost for remote resolution.
And the verification functions are fast enough.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -230,71 +215,168 @@ spec:
 
-The verification should be done in the Tekton Pipeline's admission webhook. A new configmap should be added to gate the verification
-code.
+The verification should be done in the Tekton Pipeline's reconciler after remote resolution. A new configmap should be added to gate the verification code.
```
Thanks! It should be feature-flag

Yongxuanzhang commented on teps/0091-trusted-resources.md:
````diff
@@ -230,71 +215,168 @@ spec:
+        url: "https://fulcio.sigstore.dev"
+```
+
+API (Reference from [policy-controller](https://github.com/sigstore/policy-controller)):
````
I'm proposing to leverage some APIs from policy-controller and not a whole copy. So it will have no dependency and we can adapt it to fit our needs.

Yongxuanzhang commented on teps/0091-trusted-resources.md:
```diff
@@ -112,26 +110,21 @@ be handled, or user scenarios that will be affected and must be accomodated.
+- A Pipeline Author can configure taskruns/pipelineruns to fail if a Resource cannot be verified. Or not fail but return warning.
+- Ideally, the remote registry where users store the resources should make it clear how to obtain the public key for community-provided resources
 
 ## Proposal
```
Sure!

lbernick commented:
@wlynch are you able to take another look at this PR, or are there still outstanding comments that should be resolved by Yongxuan?

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -313,12 +418,119 @@ What other approaches did you consider and why did you rule them out?  These do
+
+### Options for exposing the public keys
+
+[Public keys discovery](https://transparency.dev/application/strengthen-discovery-of-encryption-keys/) is out of scope of the first phase of this TEP and considered as part of the future work. One possible mitigation is to leverage [Fulcio](https://github.com/sigstore/fulcio) to verify the identity of signing account and issue a signed certificate. The certificate is attached to the resource and  ...
```
nit: Fulcio's great for tying keys to identities, but Rekor is what you'd want to integrate with for key discovery.

wlynch commented on teps/0091-trusted-resources.md:
```diff
@@ -313,12 +418,119 @@ What other approaches did you consider and why did you rule them out?  These do
+
+| Method | Pros | Cons |
+| -------- | ----------- | ----------- |
+| Fetch remote resources in validating Webhook | Verification fail fast in webhook | Duplicate work for Resolution, may introduce latency, and one extreme case is that the resource verified may not the the same in reconciler's resolved resource
```
Fine to leave out of this proposal for now (doesn't modify the rest of it), but I think there may be some value in still having validating webhook checks at admission for API resources without resolution (e.g. just validate that the message we received was unmodified) on top of the resolve-at-runtime behavior.

It would be useful to be able to reject unsigned resources at creation to get quick feedback on whether the signature is correct. We'd still check this again at runtime to verify the policy hasn't changed, but it can help circuit bad configs faster.

Yongxuanzhang replied:
Yeah that makes sense!

wlynch reviewed (approved):
some minor comments, but nothing worth blocking on. Feel free to follow up on these in another PR, but don't let this stop you from starting to implement!

/lgtm
